crmprtd/align.py

```diff
@@ -8,12 +8,237 @@
+        val = val.to(dst_unit).magnitude  # dest
+    except Exception:
+        raise Exception(
+            "Unable to convert source unit {} to destination unit {}".format(
+                src_unit, dst_unit)
+        )
+    return val
+
+
+# finds distance between 2 points on a sphere using lon/lat
+def haversine_formula(lat1, lon1, lat2, lon2):
+    R = 6378.137
+    dLat = lat2 * math.pi / 180 - lat1 * math.pi / 180
+    dLon = lon2 * math.pi / 180 - lon1 * math.pi / 180
+    a = math.sin(dLat/2) * math.sin(dLat/2) + math.cos(lat1 * math.pi / 180) *\
+        math.cos(lat2 * math.pi / 180) * math.sin(dLon/2) * math.sin(dLon/2)
+    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1-a))
+    d = R * c
+    return d * 1000
+
+
+def align(sesh, obs_tuple):
+    log.info('Begin alignment on row')
+
+    # place
+    log.debug('Check for network name')
+    q = sesh.query(Network).filter(Network.name == obs_tuple.network_name)
+
+    if q.count() == 0:
+        log.error('Observation cannot be used without a valid network name',
+                  extra={'network_name': obs_tuple.network_name})
+        return
+    log.debug('Found matching network name', extra={'network_name': q.first()})
+
+    log.debug('Check if station id in history')
+    q = sesh.query(History.id).join(Station).join(Network).filter(and_(
+        Network.name == obs_tuple.network_name,
+        Station.native_id == obs_tuple.station_id))
+
+    hid = None
+    if q.count() == 0:
+        log.info('No station found, creating new station',
+                 extra={'native_id': obs_tuple.station_id})
+        hid = create_station_and_history_entry(sesh, obs_tuple)
+
+    elif q.count() == 1:
+        log.debug('Matched station', extra={'history_id': q.first()})
+        hid, = q.first()
+
+    # FIXME: What happens if we cannot find an "active" station?
+    elif q.count() >= 2:
+        log.debug('Found multiple stations', extra={
+            'num_matches': q.count(), 'histories': q.all()})
+
+        if obs_tuple.lat and obs_tuple.lon:
+            matching_stns = []
+            for hid in q.first():
+                matching_stns.append(hid)
+            matching_stns = set(matching_stns)
+            close_stns = matching_stns.intersection(
+                closest_stns_within_threshold(sesh, obs_tuple.lon,
+                                              obs_tuple.lat, 800))
+            log.debug('Station set intersection',
+                      extra={'station_set': close_stns})
+
+            if len(close_stns) == 0:
+                log.debug('Station set is empty, adding new station')
+                hid = create_station_and_history_entry(sesh, obs_tuple)
+
+            elif len(close_stns) == 1:
+                log.debug('Single station in set')
+                hid = close_stns.pop()
+
+            elif len(close_stns) > 1:
+                log.debug('Multiple stations in set')
+                closest_pos = 999   # max distance should be 800m at this point
```

Comment:
IIRC `closest_stns_within_thresholds` returns its result [*ordered by distance*](https://github.com/pacificclimate/crmprtd/blob/dev/crmprtd/ec/__init__.py#L159). So if you want to select the closest, you should just be able to select the first one.